Sources/ServiceDiscovery/ServiceDiscovery.swift

```diff
@@ -12,179 +12,26 @@
-            switch self {
-            case .cancellationRequested:
-                return "cancellationRequested"
-            case .serviceDiscoveryUnavailable:
-                return "serviceDiscoveryUnavailable"
-            case .failedToMapService:
-                return "failedToMapService"
-            }
-        }
-    }
-
-    internal let type: ReasonType
-
-    public var description: String {
-        "CompletionReason.\(String(describing: self.type))"
-    }
-
-    /// Cancellation requested through `CancellationToken`.
-    public static let cancellationRequested = CompletionReason(type: .cancellationRequested)
-
-    /// Service discovery is unavailable.
-    public static let serviceDiscoveryUnavailable = CompletionReason(type: .serviceDiscoveryUnavailable)
-
-    /// A service mapping function threw an error
-    public static let failedToMapService = CompletionReason(type: .failedToMapService)
-}
-
-// MARK: - Service discovery errors
-
-/// Errors that might occur during lookup.
-public struct LookupError: Error, Equatable, CustomStringConvertible {
-    internal enum ErrorType: Equatable, CustomStringConvertible {
-        case unknownService
-        case timedOut
-
-        var description: String {
-            switch self {
-            case .unknownService:
-                return "unknownService"
-            case .timedOut:
-                return "timedOut"
-            }
-        }
-    }
-
-    internal let type: ErrorType
-
-    public var description: String {
-        "LookupError.\(String(describing: self.type))"
-    }
-
-    /// Lookup cannot be completed because the service is unknown.
-    public static let unknownService = LookupError(type: .unknownService)
-
-    /// Lookup has taken longer than allowed and thus has timed out.
-    public static let timedOut = LookupError(type: .timedOut)
-}
-
-/// General service discovery errors.
-public struct ServiceDiscoveryError: Error, Equatable, CustomStringConvertible {
-    internal enum ErrorType: Equatable, CustomStringConvertible {
-        case unavailable
-        case other(String)
-
-        var description: String {
-            switch self {
-            case .unavailable:
-                return "unavailable"
-            case .other(let detail):
-                return "other: \(detail)"
-            }
-        }
-    }
-
-    internal let type: ErrorType
-
-    public var description: String {
-        "ServiceDiscoveryError.\(String(describing: self.type))"
-    }
-
-    /// `ServiceDiscovery` instance is unavailable.
-    public static let unavailable = ServiceDiscoveryError(type: .unavailable)
-
-    public static func other(_ detail: String) -> ServiceDiscoveryError {
-        ServiceDiscoveryError(type: .other(detail))
-    }
+    /// -  Returns a ``DiscoverySequence``, which is an `AsyncSequence` and each of its items is a snapshot listing of service instances.
+    func subscribe() async throws -> DiscoverySequence
```

Comment:
One thing I wonder about is, should this method be allowed to throw. The problem is that in the `for try await` scenario, a transient issue in the service discovery service may end the for loop. This requires developers to write code like this:

```swift
while !Task.isCancelled {
  do {
    for try await let instances in discovery.subscribe() {

    }
  } catch {
    // service discovery might have problems for the moment. but we don't want service discovery to take down services that depend on it
    continue
  }
}
```

Maybe we want to consider the following: 

- a `subscribe` call can throw (which would mean that initial service discovery has failed)
- we enforce that a `DiscoverySequence` `next` call must not throw. For users to be able to observe state here, we might want to enforce a health call instead.

Reply:
that was exactly my intention, ie the throws here means the subscription setup has failed. if we think it is important to surface subscription yield issues, we can change the stream to produce Result type

Reply:
@fabianfett thoughts?

Reply:
My thinking about this is as follows:

`next()` should not show errors. Not even as a `Result`. The reason for this is `next()` will be mostly consumed by other libraries. For example a Redis library will consume the DiscoverySequence. If the next() returns a Result, what shall the library decide? I think the final decision of what to do in case of error in ServiceDiscovery is up to the application developer. Because of this, I think we might want to consider an API that helps users implement healthchecks.

What do you think about something like this?

```swift
struct HealthEvent {
    /// the number of consecutive successful updates in service discovery. Reset to zero if an error happens.
    var successfulUpdates: Int
    /// the errors of consecutive failed updates. Reset to empty array if an update is successful 
    var lastErrors: [any Error]
}

protocol ServiceDiscovery {
    associatedtype HealthSequence: AsyncSequence where HealthSequence.Element == HealthEvent
 
    func health() -> HealthSequence
}
```

This way users can implement their healthcheck endpoints using the HealthSequence, while library developers are only concerned about the success case.

Reply:
Hi!
I would add a comment regarding underlaying discovery error if you are interested.
Having a non-throwing next() and additional sequence with status will raise a question of synchronisation for those 2 streams. I suppose there are many cases when it will be enough having just a sequence of services, but often underlying service discovery can fail at any time. Failure can be recoverable (then probably the next() should not even throw something) and non-recoverable, in which case the user should decide how to proceed. Having a health sequence will require user to implement a synchronisation between those 2 streams and it will add unnecessary complexity comparing to a case when both services and errors will be in the same sequence.

Reply:
+1 we've used an AsyncStream with an enum with different payloads for data/health/status, it's a quite decent approach.

Reply:
Hmmm... I'm unsure about mixing any kind of "health" into this. Discovery doesn't say anything about health of a service; Thinking about the errors now though... I don't think it helps much to introduce some error accumulator and counter tbh